### PR TITLE
 Expose metadata as object from useProductMetadata hook

### PR DESCRIPTION
**Pull request by nathanss**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This will be useful for reading parent's metadata properties from a product variation, as an example.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

No tests are needed, I've added unit tests and also validated the behavior on an extension

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by retrofox — COMMENTED** [review]
> This hook will be useful for reading parent's metadata properties from a product variation, as an example.

Although the argument of the hooks is `parentId`, I don't see any process here about getting the parent data from the given product.
Am I missing something?
It just seems to return the metadata directly based on the given product ID.

**Review by retrofox — COMMENTED** [review]
I looked at the PR, and the API purpose isn't unclear.

The simplest usage:

```es6
const [ metadata ] = useProductMetadata()
```
Here, `metadata` will contain the metadata based on the product defined by the context.

```es6
const [ metadata ] = useProductMetadata( 'product', 10, 'product' )
```

Here, `metadata` will contain the metadata of the product based on the given ID. The third parameter is mandatory when we pass the product ID.

Could the API be more straightforward?

```es6
const [ metadata ] = useProductMetadata()
```

```es6
const [ metadata ] = useProductMetadata( { id: 10 } )
```

```es6
const [ metadata ] = useProductMetadata( { ID: 10, postType: 'product' } )
```

File: packages/js/product-editor/src/hooks/use-product-metadata.ts [hunk 1]
```diff
@@ -2,28 +2,58 @@
...
+		( select ) => {
+			let usedMetadata: Metadata< string >[];
+			if ( id ) {
+				// @ts-expect-error There are no types for this.
+				const { getEditedEntityRecord } = select( 'core' );
+				const { meta_data: parentMetadata }: Product =
+					getEditedEntityRecord( 'postType', parentPostType, id );
+				usedMetadata = parentMetadata;
+			} else {
+				usedMetadata = thisMetadata;
```
Comment (retrofox) [inline packages/js/product-editor/src/hooks/use-product-metadata.ts]: So if the ID `id` is not defined, the selector returns `thisMetadata`. Why not handle selecting the data from this selector? You could pick the post `id` either from the context by using the `useEntityId()` or from the hook param when it's defined.

**Review by retrofox — APPROVED** [review]
Thanks for addressing the suggestions 🙇‍♂️ 🚀 

just FYI, I'm going to update https://github.com/woocommerce/woocommerce/pull/43137  once this one is merged

**Comment by nathanss** [comment]
@retrofox 

Ping me when you're done and I'll review it :)

**Pull request merged**